### ISSUE TITLE
In the banding approach developed for AGN, is a spectral fit always obtained

### ISSUE THREAD

**Issue by Higginbottom** [sirocco-rt/sirocco#55]
If there are photons in a band, then Knox and Christian believe that one should always have some kind of estimate of the spectrum in the band for use in the ionization calculations.   At the 131002 meeting, it became clear that we did not know whether this was happening or not.    

So the first part of this item is to how often this is happening, and the second part is to decide what needs to be fixed.

NIck's inital notes are below:

This is a stake in the ground, mainly to remind NSH to look into this after his return from Asia.

Concerns were raised about the behaviour of spectral_estimators.c when neither a power law or exponential model could be found. It was generally thought that this should be considered a major failure of the code, perhaps even calling for a stop.

Quickly checking a benchmark AGN model, I note that the 'No suitable model in band' error occurs 12000 times.

There are 1000 cells, 30 cycles, and 10 bands, so this represents 4% of bands/cells/cycles. It definitely happens where there are many photons in a band, so I think I need to take another look at how the algorithm works, and wether it can be improved (I'm quite sure it can!!).

Basically, in the power law case, what happens is the code is trying to find the zero point of the equation

nu_mean(from model) - nu_mean(from photon flight) - for which it uses zbrent - 

which gives it the power law index, and then it uses J to compute the weight of the power law model. From memory, it often fails at this latter point, becuase even if an alpha of -15 or something gives the correct mean nu, there may be no reasonable value of W (i.e. non infinite) which gives the correct J - so the model would be thrown out.

Similar things can clearly happen in the exponential.

This definitely requires more thought and discussion, which will (I'm afraid) have to wait till November.

**Comment by Higginbottom** [comment]
Just to put a few more thoughts down:

We are not trying to fit just one observable with a two parameter model, (mean nu) we are trying to fit two - since the integrated mean intensity must be replicated. Also, we have fixed frequency bands, so even if only part of a band contains photons, we use the whole band in the fit - which can cause problems if all the photons are at one end. I suspect that this discussion will come back in the end to having some kind of intelligent banding.

**Comment by lazygun37** [comment]
I'm not sure that I really understand these points. In the power law case, for example, the mean nu should be exclusively fixed by the power law slope. (I can understand that in pathological cases that slope may be forced into unphysical regimes, but I don't think that's the main issue here.) And I can then make that power law achieve the any desired J by simply normalizing it appropriately.

Also, the issue about all the photons falling into a small fraction of a band is physically interesting (it's actually why we have an exponential model in the first place), but I don't understand why it would cause numerical trouble. If I can calculate a mean frequency, I can calculate the power law index that gives me that frequency. And I can then normalize that power law to give me the desired intensity. It'll be a terrible model, of course, but that's not really the issue.

Am I missing something here?

**Comment by Higginbottom** [comment]
So lets construct a pathological case.

We have a band running from 1e16 to 1e17 Hz,
We have a small number of photons, all concentrated right down at 1e16 Hz.
Since there are such a small number of photons, the integrated specific intensity is rather small.
The photons suggest we need alpha +10, so we now have a model

J_nu=W*nu^+10

We need to find a value for W, which means the integral of nu^+10 dnu from 1e16 to 1e17 is equal to the measured value of J.

We do this by simply doing W= J / integral, and since the integral is a known function I just do it symbolically.

Now, 1e17 ^ 11 is not a number on the mac, so the integral is NAN, so W is undefined.

So, we could limit the range of the exponent, which is what we did originally, but then you have a model that is definately wrong.

This is not an attempt to define what we should do in these cases, merely to describe how the problem arises. However, I think the solution will be something like variable banding.

**Comment by kslong** [comment]
This is exactly what we mean, Nick.  Although when you raise something to an exceptionally large power on the Mac you get NAN, in truth the number is just exceedingly large.  As long as J is not exceedingly large, the answer for W, if that is what you trying to calculate is 0.  We should not confuse machine accuracy with physics.  We want the physically right answer, regardless of the machine accuracy.  

As a general rule, when NANs start appearing in any calculation, one should figure out what causes this situation and fix it.  Sometimes, we don't do that because we think it does not matter if only occurs occasionally, but if we want a robust program this is the sort of thing we should eliminate  (when we have someone like you who is very knowledgeable of a part of the code). Sometimes, leaving things  like this around simply creates "clutter" that makes it harder to identify real problems; other times it is a symptom of a real problem.   As ionization balance is one of the most important part of Python, I don't think we should leave this problem around.

Knox

**Comment by Higginbottom** [comment]
I understand exactly what you mean. 
However, I do think that the question is more complex. If there is a real value of J in part of a band, ideally we want to model that real J for the frequency range that it exists, hence the need for dynamic banding. The current scheme will effectively give W=0 for cases where J/integral is exceedingly small, but it is not hard to envision cases where this is a bad solution.
We always knew this was an issue - we talked ourselves out of dealing with it in the summer, because it is hard, however perhaps the time has come to deal with it.

**Comment by Higginbottom** [comment]
This issue has been addressed by

a: Improvement to the power law model, to allow it to work in log space
b: Improvement ot the exp model to allow positive exponentials
c: Fix to the code which computes the mean frequency for a power law model to apply a prefactor to prevent numerical overflows
d: A rudimentary flexible banding approach where the the minimum and maximum frequency photon seen in each spectral band are logged. These are used as the limits for the band, unless the lower/highest frequency seen is 'close' to a band edge, where close is defined as bandmax-bandmin / sqrt(N) where N is the number of photons seen.

This band is then used for integrations, and also to decide if a spectral model is available to compute the radiation density required for induced compton heating.
